src/rebar_prv_common_test.erl

```diff
@@ -37,54 +38,352 @@ init(State) ->
+    ?CONSOLE("All ~p tests passed.", [Passed]);
+format_result({Passed, Failed, Skipped}) ->
+    Format = [format_failed(Failed), format_skipped(Skipped),
+              format_passed(Passed)],
+    ?CONSOLE("~s", [Format]).
+
+format_failed(0) ->
+    [];
+format_failed(Failed) ->
+    io_lib:format("Failed ~p tests. ", [Failed]).
+
+format_passed(Passed) ->
+    io_lib:format("Passed ~p tests. ", [Passed]).
+
+format_skipped({0, 0}) ->
+    [];
+format_skipped({User, Auto}) ->
+    io_lib:format("Skipped ~p (~p, ~p) tests. ", [User+Auto, User, Auto]).
+
+test_state(State) ->
+    TestOpts = case rebar_state:get(State, ct_compile_opts, []) of
+        []    -> [];
+        Opts  -> [{erl_opts, Opts}]
+    end,
+    [first_files(State)|TestOpts].
+
+first_files(State) ->
+    CTFirst = rebar_state:get(State, ct_first_files, []),
+    {erl_first_files, CTFirst}.
+
+setup_ct(State) ->
+    Opts = resolve_ct_opts(State),
+    Opts1 = discover_tests(State, Opts),
+    copy_and_compile_tests(State, Opts1).
+
+resolve_ct_opts(State) ->
+    {RawOpts, _} = rebar_state:command_parsed_args(State),
+    CmdOpts = transform_opts(RawOpts),
+    CfgOpts = rebar_state:get(State, ct_opts, []),
+    Merged = lists:ukeymerge(1,
+                             lists:ukeysort(1, CmdOpts),
+                             lists:ukeysort(1, CfgOpts)),
+    %% make sure `dir` and/or `suite` from command line go in as
+    %%  a pair overriding both `dir` and `suite` from config if
+    %%  they exist
+    case {proplists:get_value(suite, CmdOpts), proplists:get_value(dir, CmdOpts)} of
+        {undefined, undefined} -> Merged;
+        {_Suite, undefined}    -> lists:keydelete(dir, 1, Merged);
+        {undefined, _Dir}      -> lists:keydelete(suite, 1, Merged);
+        {_Suite, _Dir}         -> Merged
+    end.
+
+discover_tests(State, Opts) ->
+    case proplists:get_value(spec, Opts) of
+        undefined -> discover_dirs_and_suites(State, Opts);
+        TestSpec  -> discover_testspec(TestSpec, Opts)
+    end.
+
+discover_dirs_and_suites(State, Opts) ->
+    case {proplists:get_value(dir, Opts), proplists:get_value(suite, Opts)} of
+        %% no dirs or suites defined, try using `$APP/test` and `$ROOT/test`
+        %%  as suites
+        {undefined, undefined} -> test_dirs(State, Opts);
+        %% no dirs defined
+        {undefined, _} -> Opts;
+        %% no suites defined
+        {_, undefined} -> Opts;
+        %% a single dir defined, this is ok
+        {Dirs, Suites} when is_integer(hd(Dirs)), is_list(Suites) -> Opts;
+        %% still a single dir defined, adjust to make acceptable to ct
+        {[Dir], Suites} when is_integer(hd(Dir)), is_list(Suites) ->
+            [{dir, Dir}|lists:keydelete(dir, 1, Opts)];
+        %% multiple dirs and suites, error now to simplify later steps
+        {_, _} -> erlang:error({multiple_dirs_and_suites, Opts})
+    end.
+
+discover_testspec(_TestSpec, Opts) ->
+    lists:keydelete(auto_compile, 1, Opts).
```

Comment:
Is this function incomplete or there is actually nothing to discover in testspecs? If so, it might be worth a comment.